streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+import org.apache.kafka.streams.kstream.KeyValueMapper;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.apache.kafka.streams.kstream.Reducer;
+import org.apache.kafka.streams.kstream.TimeWindows;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.WindowedDeserializer;
+import org.apache.kafka.streams.kstream.internals.WindowedSerializer;
+
+import java.util.Properties;
+
+/**
+ * Demonstrates, using the high-level KStream DSL, how to implement an IoT demo application
+ * which ingests temperature value processing the maximum value in the latest 5 seconds sending
+ * a new message if it exceeds 20
+ *
+ * In this example, the input stream reads from a topic named "temperature", where the values of messages
+ * represent temperature values; using a 5 seconds "tumbling" window, the maximum value is processed and
+ * sent to a topic named "max" if it exceeds 20
+ *
+ * Before running this example you must create the input topic and the output topic (e.g. via
+ * bin/kafka-topics.sh --create ...), and write some data to the input topic (e.g. via
+ * bin/kafka-console-producer.sh). Otherwise you won't see any data arriving in the output topic.
+ */
+public class TemperatureDemo {
+
+    public static void main(String[] args) throws Exception {
+
+        Properties props = new Properties();
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "streams-temperature");
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        props.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        props.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+
+        KStreamBuilder builder = new KStreamBuilder();
+
+        KStream<String, String> source = builder.stream("temperature");
+
+        KStream<Windowed<String>, String> max = source
+                .map(new KeyValueMapper<String, String, KeyValue<String, String>>() {
+                    @Override
+                    public KeyValue<String, String> apply(String key, String value) {
+                        return new KeyValue<>("temp", value);
+                    }
+                })
+                .groupByKey()
+                .reduce(new Reducer<String>() {
+                    @Override
+                    public String apply(String value1, String value2) {
+                        if (Integer.parseInt(value1) > Integer.parseInt(value2))
+                            return value1;
+                        else
+                            return value2;
+                    }
+                }, TimeWindows.of(5000))
+                .toStream()
```

Comment:
@ppatierno Okay, I think I finally get your points now. And that's the beauty of written communications :) What I had originally is that

1. Hmm, `toStreams` seems not needed since `KTable` has a `filter` operator as well.
2. Then I got a reply from you that seems to say `KTable#filter` seems incorrect that it prints unexpected outputs and that is why I have to use `KTable.toStream.filter()`.
3. Then I replied again saying well, that output is actually expected due to the `KTable#filter` semantics.

Now regarding to whether that semantics fit well with the app logic that depends on how we interpret the app. Let's say if your purpose is building an online monitoring system that alerts whenever degree is beyond threshold 20 on a windowing basis of 5 minutes, and then let's say there a degree readings of

```
10, 25, 15
```

For a specific window, does it need to trigger an alert, since it "has ever exceeded the threshold", or not since "its final value within that degree is below the threshold":

1. In the former case, I think a windowing operation would be meaningless since we would have to trigger an alert (i.e. write that into the final output topic so that it can be read and alert triggered) whenever there is a value higher than threshold; but that seems to be your application logic, with `toStream.filter` your output messages are `25` only. But as I said you would not need a windowed reduce function at all but just `stream.filter().to()`.

2. In the latter case, then a windowing operation does make sense, but then you should not alert on `25` since that is not the final value within the window, and with `table.filter()` your output would be `null` indicating that for that window the final value does not actually exceed the threshold and hence we do not need to alert.

Does that make sense to you?